Phase 2 HLT menu targetting 7.5×10³⁴ cm⁻²s⁻¹ [11.1.x]

fwyzard opened the pull request:
#### PR description:

Implement the Phase-2 HLT menu used for the HLT TDR.

HLT_75e33_cfg.py implements a self-contained configuration:
  - selects the required input collections for running on a re-emulated L1 sample;
  - uses the global tag 111X_mcRun4_realistic_T15_v5;
  - configures the MessageLogger, DQMStore, FastTimerService and ThroughputService.

HLT_75e33_cff.py implements a configuration fragment for use wih cmsDriver. To avoid redefinition errors, most of the L1T modules and the offline beamspot module are dropped from this configuration.

The HLT configuration is based on
  - `CMSSW_11_1_8`
  - global tag [`111X_mcRun4_realistic_T15_v5`](https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/gts/111X_mcRun4_realistic_T15_v5)
  - the Tracking v6.1 baseline configuration
  - the TICL v3 configuration

and implement the baseline Phase-2 configuration for Jets/MET, b-tagging, Muons and E/Gamma.
Taus are not included.

It includes "open" or "MC" HLT paths for running the reconstruction without any (pre)selection applied, and a partial re-emulation of the Level-1 Trigger, which simulates a simplified L1T menu using EDFilters and Paths.

For Jets/MET and B-tagging, the exact results obtained for the TDR should be reproducible with the commit 19f3c18687c0 (and likely also with 93a4a572b336, which is just a technical fix), which still use "offline" muons for the Particle Flow reconstruction.
The following commits implement the use of HLT muons in the Particle Flow reconstruction, affecting the Jets/MET (and likely B-tagging) performance in a non-negligible way; for more details see [this presentation](https://indico.cern.ch/event/1035816/contributions/4350085/subcontributions/337175/attachments/2242877/3803185/210511_hltupg_val32903.pdf).

For Muons, E/Gamma, and overall timing measurements, the exact results obtained in the TDR should be reproducible using the full PR.


#### PR validation:

The configuration in `HLTrigger/Configuration/python/HLT_75e33_cfg.py` is as close as it can be to the one used for the HLT TDR.

The configuration in `HLTrigger/Configuration/python/HLT_75e33_cff.py` is almost identical, and was validated by running:
```
$ cmsDriver.py step2  --conditions auto:phase2_realistic_T15 -s HLT:75e33 --geometry Extended2026D49 --era Phase2C9 --eventcontent FEVTDEBUGHLT --datatier GEN-SIM-DIGI-RAW -n 10 --filein /store/mc/Phase2HLTTDRSummer20ReRECOMiniAOD/MinBias_TuneCP5_14TeV-pythia8/SKIM/PU200_111X_mcRun4_realistic_T15_v1-v1/000001/2cf57ee0-f543-4718-a2f1-db8b9baeac87.root --processName=HLTX -n 20 --nThreads 8
...
(skip various initialisation messges)
Begin processing the 1st record. Run 1, Event 162234, LumiSection 917 on stream 5 at 19-Sep-2021 11:06:02.253 CEST
%MSG-w TwoTrackMinimumDistance:   InclusiveCandidateVertexFinder:hltDeepInclusiveVertexFinderPF  19-Sep-2021 11:06:27 CEST Run: 1 Event: 162234
comparing track with itself!
%MSG
...
Begin processing the 14th record. Run 1, Event 519298, LumiSection 2934 on stream 6 at 19-Sep-2021 11:08:53.867 CEST
Begin processing the 15th record. Run 1, Event 519318, LumiSection 2934 on stream 7 at 19-Sep-2021 11:08:53.881 CEST
Begin processing the 16th record. Run 1, Event 169611, LumiSection 959 on stream 1 at 19-Sep-2021 11:09:18.739 CEST
Begin processing the 17th record. Run 1, Event 127902, LumiSection 723 on stream 7 at 19-Sep-2021 11:09:34.448 CEST
Begin processing the 18th record. Run 1, Event 127910, LumiSection 723 on stream 6 at 19-Sep-2021 11:09:34.466 CEST
Begin processing the 19th record. Run 1, Event 127969, LumiSection 723 on stream 3 at 19-Sep-2021 11:09:34.480 CEST
Begin processing the 20th record. Run 1, Event 526297, LumiSection 2974 on stream 0 at 19-Sep-2021 11:10:09.478 CEST
19-Sep-2021 11:10:24 CEST  Closed file file:///gpu_data/store/mc/Phase2HLTTDRSummer20ReRECOMiniAOD/MinBias_TuneCP5_14TeV-pythia8/SKIM/PU200_111X_mcRun4_realistic_T15_v1-v1/000001/2cf57ee0-f543-4718-a2f1-db8b9baeac87.root

=============================================

MessageLogger Summary

 type     category        sev    module        subroutine        count    total
 ---- -------------------- -- ---------------- ----------------  -----    -----
    1 TwoTrackMinimumDista -w InclusiveCandida                      72       72
    2 fileAction           -s file_close                             1        1
    3 fileAction           -s file_open                              2        2

 type    category    Examples: run/evt        run/evt          run/evt
 ---- -------------------- ---------------- ---------------- ----------------
    1 TwoTrackMinimumDistance 1/162234         1/162234         1/512433
    2 fileAction           PostGlobalEndRun                  
    3 fileAction           pre-events       pre-events       

Severity    # Occurrences   Total Occurrences
--------    -------------   -----------------
Warning                72                  72
System                  3                   3
```

missirol committed:
Add the baseline Phase 2 HLT configuration for Jets/MET

This configuration is based on
  - CMSSW 11.1.8
  - global tag 111X_mcRun4_realistic_T15_v5
  - the Tracking v6.1 baseline configuration
  - the TICL v3 configuration

SWuchterl committed:
Add the baseline Phase 2 HLT configuration for b-tagging

The b-tagging paths should be consistent with the Jets/MET configuration.

cmsbuild commented:
A new Pull Request was created by  @fwyzard (Andrea Bocci) for CMSSW_11_1_X.

It involves the following packages:

HLTrigger/Phase2


The following packages do not have a category, yet:

HLTrigger/Phase2
Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories_map.py to assign category

@cmsbuild can you please review it and eventually sign? Thanks.
@Martin-Grunewald this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

fwyzard commented:
@Sam-Harper @trtomei FYI

missirol commented:
I'm copying here for reference a quick exchange with Andrea, regarding a few small follow-ups to this PR, to be done in the next days.

--------------------

(quoted text from me, answers from Andrea)

> * process.options contains various strings that reference RECO2, so they’d probably have no effect in this config; should we change this to HLTX, or remove them altogether (I don’t know what the canDeleteEarly option actually does)?

Yes, this needs to be cleaned.

> * a TFileService module slipped in when the BTV part was integrated; I guess it is better to remove it.

Yes.

> * I made a mistake, because I effectively gave you a path with “PFMET+MHT”, but in the TDR we plan to use “PFMET Type1” rather than (Raw) PFMET. How do you prefer to proceed in this case? I can update the recipe in my user package (and you re-export), or I can make a PR to your branch with the necessary changes (it is a relatively simple change, see [here](https://github.com/missirol/JMETriggerAnalysis/commit/dec03d201a373a5d90f60f0f8766f0ba65362350)).

Make a PR to my branch, please.

> * in HLT_75e33_cfg.py the BTV paths are missing (while they are present in HLT_75e33.py); you probably did it intentionally, though, because if I include them I get an error (some cms-data files called by BTV are missing).

No, I just forgot to include it in the update :-(

> * right now we have a MC path (called “MC_JME”) which is not really a path, but just a sequence of a lot of different collections, many of which we don’t really use (e.g. AK4PFJets, AK4PFCHSJets, and AK8* jets). I had removed it from the JME config, but it somehow reappeared when you integrated the BTV part. What do you prefer to do with it? I would say it is not really needed (we could at least remove the collections that are not used in any of the actual triggers).

I would keep it and make it optional, so people can use it to make ntuples without a priori cuts.
Of course, cleaning it up is appreciated :-)

> * one thing that we will have to change at some point is the HLT thresholds (for JME), because right now they are simply taken from 2018 for that config (e.g. Jet500, HT1050, MET120_MHT120).

OK, sure.

Sam-Harper commented:
Thanks Andrea. 

I do have a small question on technical nature of the menu structure.  The issue is that the menu loads manually in all objects. For E/gamma we took a different approach
https://gitlab.cern.ch/sharper/EgHLTPhase2/-/blob/master/Menus/egammaBasicMenu.py

which is where the menu just loads the path and the path is assumed to be able to load everything it needs in. Is there an issue with doing it this way? 

For esproducers, esources and psets, this is of course difficult to do (although thinking about it there is no reason why modules requiring external psets can not just import them so that when they imported the pset is there) and I default to loading all available. While I admit this is a not a great solution, having them all loaded in the menu one by one also doesnt seem great it will likely become tedious to maintain and the ultimate effect will likely be just loading everything anyways.

fwyzard commented:
Sure.

Please make a PR with the necessary changes to edmConfigSplit .

missirol committed:
Update the Jets/MET configuration to match the AR and draft TDR

Update the Jets/MET configuration:
  - replace PFPuppiMET with PFPuppiMETTypeOne
  - update the Jets/MET thresholds to match the AR and draft TDR
  - remove the PF CHS reconstruction

fwyzard commented:
@missirol 

> * process.options contains various strings that reference RECO2, so they’d probably have no effect in this config; should we change this to HLTX, or remove them altogether (I don’t know what the canDeleteEarly option actually does)?

`canDeleteEarly` tells the framework that there are no dependencies (`edm::Ref`, `edm::Ptr`, raw pointers, *etc.*) on these collections, and that they can be deleted as soon as there are no other modules that depend on it that need to be run.

It'll probably make sense to drop these entirely, but for the time being I've changed them to `HLTX`.


> * a TFileService module slipped in when the BTV part was integrated; I guess it is better to remove it.

Done.

> * I made a mistake, because I effectively gave you a path with “PFMET+MHT”, but in the TDR we plan to use “PFMET Type1” rather than (Raw) PFMET. How do you prefer to proceed in this case? I can update the recipe in my user package (and you re-export), or I can make a PR to your branch with the necessary changes (it is a relatively simple change, see [here](https://github.com/missirol/JMETriggerAnalysis/compare/3a72b..5e778)).

Thanks for having updated your setup - these changes should now be included here as well.

> * in HLT_75e33_cfg.py the BTV paths are missing (while they are present in HLT_75e33.py); you probably did it intentionally, though, because if I include them I get an error (some cms-data files called by BTV are missing).

I've regenerated the full dump now.

> * one thing that we will have to change at some point is the HLT thresholds (for JME), because right now they are simply taken from 2018 for that config (e.g. Jet500, HT1050, MET120_MHT120).

Let me know if you have an update on this.

missirol commented:
Thanks, @fwyzard.

>> * one thing that we will have to change at some point is the HLT thresholds (for JME), because right now they are simply taken from 2018 for that config (e.g. Jet500, HT1050, MET120_MHT120).
>
> Let me know if you have an update on this.

There is no particular update on this; these HLT thresholds are effectively placeholders at the moment (I could replace them with the thresholds reported at the AR, though). My thinking was to update them once we have the final rate estimates that will be used in the TDR (but let me know if you have something different in mind).

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

Martin-Grunewald commented:
@fwyzard 
Indeed I prefer putting them under HLTrigger/Configuration with the directory structure you indicate above!

cmsbuild commented:
Pull request #32903 was updated. @Martin-Grunewald, @cmsbuild, @fwyzard can you please check and sign again.

missirol commented:
please test

SWuchterl committed:
Update the B-tagging configuration

Remove unused sequences and modules:
  - BProbability for Puppi and PF jets, and their corresponding submodules;
  - HLTObjects and L1Objects;
  - DeepCSV and DeepJet paths for PFCHS jets, and their corresponding submodules.

Merge all noFilter paths into MC_BTV.

Various renames:
  - remove _v1 tags from path names;
  - rename L1T filters to match the implementation;
  - rename hltDoublePFPuppiJets128Eta2p3MaxDeta1p6_cfi to hltDoublePFPuppiJets128Eta2p4MaxDeta1p6_cfi because of wrong eta cut in name.

cmsbuild commented:
Pull request #32903 was updated. @Martin-Grunewald, @cmsbuild, @fwyzard can you please check and sign again.

khaosmos93 committed:
Implement the Phase 2 HLT configuration for Muons

At the same time:
  - remove unused PSets and modules
  - clean up whitespcaes
  - clean up the imports in the main configuration file

Co-authored-by: Thiago Tomei <thiago.tomei@cern.ch>

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

fwyzard commented:
@cms-sw/hgcal-dpg-l2 FYI

fwyzard commented:
@missirol @SWuchterl can you review the impact on the Jets/MET and b-tag paths ?

fwyzard commented:
@khaosmos93 could you review the implementation of the Muons paths ?

missirol commented:
> @missirol [..] can you review the impact on the Jets/MET [..] paths ?

Hi @fwyzard , a quick update:

* a small technical fix is needed to test the configuration, see https://github.com/fwyzard/cmssw/pull/10;

* the latest commit in this PR ("Integrate muons in the Phase 2 HLT menu") causes changes in the outputs for Jets and MET (checked on only 100 events); i will try to understand what the reason is.

fwyzard commented:
I think some changes are expected, because the muons enter the PF reconstruction. Hopefully the changes are not too large...

missirol commented:
Okay, I see, thanks. Will try to quantify how large the differences are.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

Sam-Harper committed:
Implement the Phase 2 HLT configuration for E/Gamma

Add E/Gamma-related products to the input event content, and implement
the L1 E/Gamma reconstruction.

Implement the E/Gamma HLT reconstruction and trigger paths.

fwyzard committed:
Update all trigger paths with the HLTBeginSequence and HLTEndSequence

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

fwyzard committed:
Update the L1T simulation

Update the L1T simulation:
  - update to the latest official recipe
  - enable L1T PF candidates
  - implement L1T NN taus

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

fwyzard committed:
Clean up the inputs, modules, tasks, sequences and paths

General configuration clean up:
  - remove duplicate or unused modules;
  - remove empty tasks;
  - remove unused or duplicate paths.

Remove unused L1T simulation paths.

Update the list of collections read form the input files.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

missirol commented on HLTrigger/Phase2/python/HLT_75e33/modules/L1TkMuons_cfi.py:
```diff
@@ -0,0 +1,49 @@
+    L1EMTFInputTag = cms.InputTag("simEmtfDigis","EMTF"),
+    L1EMTFTrackCollectionInputTag = cms.InputTag("simEmtfDigis"),
+    L1OMTFInputTag = cms.InputTag("simOmtfDigis","OMTF"),
+    L1TrackInputTag = cms.InputTag("TTTracksFromTrackletEmulation","Level1TTTracks","RECO"),
```
@fwyzard I think this process label could be removed, too.

fwyzard replied:
Thanks, should be fixed now.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild can you please check and sign again.

fwyzard commented on HLTrigger/Phase2/python/HLT_75e33/modules/l1tDoublePFPuppiJets112offMaxDeta1p6_cfi.py:
```diff
@@ -14,8 +14,8 @@
-    originTag1 = cms.VInputTag("l1tSlwPFPuppiJetsCorrected::Phase1L1TJetFromPfCandidates"),
-    originTag2 = cms.VInputTag("l1tSlwPFPuppiJetsCorrected::Phase1L1TJetFromPfCandidates"),
+    originTag1 = cms.VInputTag(cms.InputTag("l1tSlwPFPuppiJetsCorrected","Phase1L1TJetFromPfCandidates")),
+    originTag2 = cms.VInputTag(cms.InputTag("l1tSlwPFPuppiJetsCorrected","Phase1L1TJetFromPfCandidates")),
```
@SWuchterl looks like there was a typo here, `::` instead of a single `:`

fwyzard committed:
Add a configuration fragment for use wih cmsDriver

Delete L1T and beam spot modules from the HLT configuration to avoid redefinition errors
when running in the same process as the L1T simulation.

cmsbuild commented:
Pull request #32903 was updated. @Martin-Grunewald, @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @Martin-Grunewald, @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @Martin-Grunewald, @cmsbuild can you please check and sign again.

fwyzard commented:
@missirol @khaosmos93 @Sam-Harper @SWuchterl @trtomei thanks again for the contribution.

fwyzard commented:
@Martin-Grunewald @missirol I would suggest pushing this PR to 11.1.x (even if we don't ask for a new release) to have the configuration in release for reference.

It can also be used as the starting point for a 12.1.x PR, of course.

fwyzard commented:
please test

fwyzard commented:
(to make the bot happy, even though there is nothing in the release that will exercise the new menu)

missirol commented:
Thanks, @fwyzard.

> I would suggest pushing this PR to 11.1.x (even if we don't ask for a new release) to have the configuration in release for reference.

I would agree, and I'd also think it's useful to ask for a new release once this PR is merged; this way we'd have a clear reference to reproduce the TDR results [*]. You, Martin and others can give a more informed opinion, in case there are any downsides in asking for a new release.

On a technical note, I see that `c1a6acc` in this PR is a copy of #33205. Is it maybe better to rebase on top of the latter?

Finally, could you remind me what was the plan/status concerning the integration of tau triggers in the Phase-2 menu?

[*] A note, just for completeness: this PR does not reproduce exactly the results in the TDR for JME (and I guess also BTV), because it already implements the use of HLT muons in the PFlow modules (for the JME results in the TDR, PFlow used 'offline' muons, as JME and MUO developments were proceeding somewhat in parallel; [more info here](https://indico.cern.ch/event/1035816/contributions/4350085/subcontributions/337175/attachments/2242877/3803185/210511_hltupg_val32903.pdf)). Ideally, we could have maybe squashed these updates having 1 intermediate commit before the merge of HLT-muons and PFlow, but I forgot to make this point earlier.

cc: @beaucero

cmsbuild commented:
Pull request #32903 was updated. @Martin-Grunewald, @cmsbuild can you please check and sign again.

fwyzard commented:
> I would agree, and I'd also think it's useful to ask for a new release once this PR is merged; this way we'd have a clear reference to reproduce the TDR results [*]. You, Martin and others can give a more informed opinion, in case there are any downsides in asking for a new release.

Agreed, having a release would make things easier to reproduce and reuse in the future.

> On a technical note, I see that `c1a6acc` in this PR is a copy of #33205. Is it maybe better to rebase on top of the latter?

Done, thanks for pointing it out.

> Finally, could you remind me what was the plan/status concerning the integration of tau triggers in the Phase-2 menu?

Ehr... I _think_ @trtomei gave it a try a couple of months ago, but it didn't work out as expected ?
I'm not sure what the problem may have been, nor what is the best way forward (trying again the integration in 11.1.x, or moving to 12.1.x, or some other idea).

> [*] A note, just for completeness: this PR does not reproduce exactly the results in the TDR for JME (and I guess also BTV), because it already implements the use of HLT muons in the PFlow modules (for the JME results in the TDR, PFlow used 'offline' muons, as JME and MUO developments were proceeding somewhat in parallel; [more info here](https://indico.cern.ch/event/1035816/contributions/4350085/subcontributions/337175/attachments/2242877/3803185/210511_hltupg_val32903.pdf)). Ideally, we could have maybe squashed these updates having 1 intermediate commit before the merge of HLT-muons and PFlow, but I forgot to make this point earlier.

The reason for squashing all the commits is mostly technical: as I moved everything under `HLTrigger/Configueration`, I wanted to avoid adding-deleting-readding all the files - otherwise I would have kept the full history :-(

If there is enough interest, I can instead try to replay the whole history in the final directory - but out of hand I don't know how hard it's going to be.

fwyzard committed:
Add the 7.5e34 HLT menu to the Phase-2 workflows using the D49 geometry

cmsbuild commented:
Pull request #32903 was updated. @Martin-Grunewald, @cmsbuild can you please check and sign again.

fwyzard commented:
OK, I've recreated the relevant part of the history instead of squashing it.

fwyzard commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-741dce/18745/summary.html
**COMMIT**: c8fde920cdf9b32a53eb8c1455f6af10b79c7d49
**CMSSW**: CMSSW_11_1_X_2021-09-19-0000/slc7_amd64_gcc820
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32903/18745/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2784828
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2784777
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 31 edm output root files, 36 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_1_X_2021-09-19-0000+741dce/45589/triggerResults): no differences found

missirol commented:
> OK, I've recreated the relevant part of the history instead of squashing it.

Thanks, Andrea.

So, for reference, the JME (and BTV) results in the TDR can be reproduced with 499aa31b8692bfa3e85042e97928a40465c8815f (modulo the fact that one needs the fix in d5a8a5416be229076342556f5cc60e71b812ef5d).

fwyzard commented:
OK, then let me see if I can clean up the history and keep only the interesting intermediate steps.

Martin-Grunewald commented:
@fwyzard 
Is there a relval or other test exercising the new menu?

fwyzard commented:
> @fwyzard
> Is there a relval or other test exercising the new menu?

Not yet; I've tried adding it as the standard HLT step for the Phase-2 relvals, but it didn't work out of the box.

I suspect it may be due to the Phase-2 L1 simulation in 11.0.x/11.1.x not being complete (actually, not being there at all).
I'm not sure if it's worth investigating it for 11.1.x, or do it properly for 12.1.x, which hopefully includes at least some of it.

fwyzard commented:
> So, for reference, the JME (and BTV) results in the TDR can be reproduced with [499aa31](https://github.com/cms-sw/cmssw/commit/499aa31b8692bfa3e85042e97928a40465c8815f) (modulo the fact that one needs the fix in [d5a8a54](https://github.com/cms-sw/cmssw/commit/d5a8a5416be229076342556f5cc60e71b812ef5d)).

Going through the commits _after_ 499aa31b8692bfa3e85042e97928a40465c8815f, I hope that the TDR results included also the fix from e1ed92b6cf61715ac3e58d4ae0489c8793f532c1 (or that it is irrelevant) ?

missirol commented:
> > So, for reference, the JME (and BTV) results in the TDR can be reproduced with [499aa31](https://github.com/cms-sw/cmssw/commit/499aa31b8692bfa3e85042e97928a40465c8815f) (modulo the fact that one needs the fix in [d5a8a54](https://github.com/cms-sw/cmssw/commit/d5a8a5416be229076342556f5cc60e71b812ef5d)).
> 
> Going through the commits _after_ [499aa31](https://github.com/cms-sw/cmssw/commit/499aa31b8692bfa3e85042e97928a40465c8815f), I hope that the TDR results included also the fix from [e1ed92b](https://github.com/cms-sw/cmssw/commit/e1ed92b6cf61715ac3e58d4ae0489c8793f532c1) , or that it is irrelevant (_e.g._ because we run the L1T simulation in a separate job).
> 
> Still, I would prefer to include it in the intermediate "TDR version" of the history.

Indeed, I think this is also needed. It's for a L1T seed used by BTV. @SWuchterl can you confirm you had this fix in for the BTV results of the TDR?

SWuchterl commented:
Thanks for pointing this out. Results we have produced, i.e., performance and rate results, are not affected. I think if it was also fixed for the central runs by the commit in this PR, or fixed L1T and HLT steps are separated as @fwyzard said, everything should be fine.

cmsbuild commented:
Pull request #32903 was updated. @Martin-Grunewald, @cmsbuild can you please check and sign again.

fwyzard commented:
> Thanks for pointing this out. Results we have produced, i.e., performance and rate results, are not affected. I think if it was also fixed for the central runs by the commit in this PR, or fixed L1T and HLT steps are separated as @fwyzard said, everything should be fine.

OK, I'll add it in for the "TDR version".

cmsbuild commented:
Pull request #32903 was updated. @Martin-Grunewald, @missirol, @cmsbuild can you please check and sign again.

fwyzard commented:
OK, this is all the history clean up I plan to do.

cmsbuild commented:
Pull request #32903 was updated. @Martin-Grunewald, @missirol, @cmsbuild can you please check and sign again.

cmsbuild commented:
Pull request #32903 was updated. @Martin-Grunewald, @missirol, @cmsbuild can you please check and sign again.

fwyzard commented:
And now with a more reasonable authorship.

cmsbuild commented:
Pull request #32903 was updated. @Martin-Grunewald, @missirol, @cmsbuild can you please check and sign again.

missirol commented:
Thanks for the extra work, @fwyzard. Looks good to me.

The commit to reproduce the JME and BTV results in the TDR is now 93a4a57.

Martin-Grunewald commented:
please test

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild, @jordan-martins, @missirol, @bbilin, @wajidalikhan, @Martin-Grunewald, @AdrianoDee, @srimanob, @kskovpen can you please check and sign again.

Martin-Grunewald commented:
please test

fwyzard commented:
A possible improvement would be to load the duplicate modules from the offline configuration files; hopefully this makes CMSSW realise that they are the same, and ignore the duplicate load.

fwyzard commented:
IF you want to try the cmsDriver test, I'm using
```bash
runTheMatrix.py -w relval_2026 -l 23234.21 -t 16
```

cmsbuild commented:
-1

**Failed Tests**: RelVals RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-741dce/18754/summary.html
**COMMIT**: 9bee864f2cf3da1ccc9f51e7d918a779a2e82f13
**CMSSW**: CMSSW_11_1_X_2021-09-19-0000/slc7_amd64_gcc820
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32903/18754/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals

- [20034.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-741dce/18754/runTheMatrix-results/20034.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D35_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D35+RecoFullGlobal_2026D35+HARVESTFullGlobal_2026D35)
```
----- Begin Fatal Exception 20-Sep-2021 14:52:03 CEST-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'L1T_DoubleNNTau52'
   [2] Calling method for module L1TPFTauFilter/'hltL1DoubleNNTau52'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: std::vector<l1t::PFTau>
Looking for module label: l1NNTauProducerPuppi
Looking for productInstanceName: L1PFTausNN

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------
```
- [21234.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-741dce/18754/runTheMatrix-results/21234.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D44_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D44+RecoFullGlobal_2026D44+HARVESTFullGlobal_2026D44)
```
----- Begin Fatal Exception 20-Sep-2021 14:56:44 CEST-----------------------
An exception of category 'BadConfiguration' occurred while
   [0] Processing  stream begin Run run: 1 stream: 0
   [1] Calling method for module MixingModule/'mix'
Exception Message:
HGCDigitizer is not producing EE, FH, or BH digis!
----- End Fatal Exception -------------------------------------------------
```


## RelVals-INPUT

- [20034.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-741dce/18754/runTheMatrixINPUT-results/20034.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D35_GenSimHLBeamSpotFull14INPUT+DigiFullTrigger_2026D35+RecoFullGlobal_2026D35+HARVESTFullGlobal_2026D35)```20034.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D35_GenSimHLBeamSpotFull14INPUT+DigiFullTrigger_2026D35+RecoFullGlobal_2026D35+HARVESTFullGlobal_2026D35/step2_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D35_GenSimHLBeamSpotFull14INPUT+DigiFullTrigger_2026D35+RecoFullGlobal_2026D35+HARVESTFullGlobal_2026D35.log```

fwyzard commented:
please test

cmsbuild commented:
Pull request #32903 was updated. @jordan-martins, @missirol, @bbilin, @wajidalikhan, @Martin-Grunewald, @AdrianoDee, @srimanob, @kskovpen can you please check and sign again.

fwyzard commented:
looks like the L1T or the HLT will not work with some geometries:
  - C4: HGCal v9 + Phase2 HCAL and EB
  - F3/F5: changes due to the HFNose
  - I12: new ETL layout from MTD TDR

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-741dce/18766/summary.html
**COMMIT**: 4e6142d6a78bbbff84e4618a210494eb9fa6cace
**CMSSW**: CMSSW_11_1_X_2021-09-19-0000/slc7_amd64_gcc820
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32903/18766/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 2678 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2784828
 - DQMHistoTests: Total failures: 26708
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2758070
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 31 edm output root files, 36 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_1_X_2021-09-19-0000+741dce/45602/triggerResults): no differences found

missirol commented:
It seems that differences in [wf 23234.0](https://tinyurl.com/yel2fqaq) are to be expected.

The DQM modules showing differences get filled based on the names of the paths in `TriggerResults::HLT` (which have now changed for that Phase-2 wf), see

https://github.com/cms-sw/cmssw/blob/8729a0bf88199e4c61f7d3bba9464e0d44af082b/DQM/Physics/python/qcdPhotonsDQM_cfi.py#L6
https://github.com/cms-sw/cmssw/blob/8729a0bf88199e4c61f7d3bba9464e0d44af082b/DQM/Physics/python/ExoticaDQM_cfi.py#L9

Martin-Grunewald commented:
How to proceed? Have just one compatible/working HLT-PhaseII test, not all of them?

fwyzard commented:
> Have just one compatible/working HLT-PhaseII test, not all of them?

I'm not sure I understand the question.

Martin-Grunewald commented:
At the moment in this PR you switched all Phase-II HLT workflows to use the Phase-II HLT menu (@relval2026 is pointing to it).
The suggestion would be to only have one (or a few) with compatible setup/geometry using the actual menu. with the rest reverting to Fake2.

fwyzard commented:
Actually, that is what I have done in the last commit: https://github.com/cms-sw/cmssw/pull/32903/commits/4e6142d6a78bbbff84e4618a210494eb9fa6cace .

Martin-Grunewald commented:
Ah great!

srimanob commented:
See that D46 can run, which is great as it is the only one remain in 12_1. 
What is the plan of integration? It seems fixings in the master may not needed to port back to 11_1.

fwyzard commented:
I would suggest making no further changes to the 11.1.x branch, and make any further developments, refactoring, etc. only for master.

missirol commented:
> I would suggest making no further changes to the 11.1.x branch, and make any further developments, refactoring, etc. only for master.

Fwiw, I agree with this.

missirol commented:
Looking at the PR tests, it seems that the HLT ran as intended in the relevant Phase-2 wfs. A couple of notes:

* the [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_1_X_2021-09-19-0000+741dce/45602/triggerResults/) comparisons are misleading at first glance ("no differences found"). It seems this happens b/c `hltDiff` only checks common paths between old and new files: "Warning: old and new TriggerResults come from different HLT menus. Only the common 4 triggers are compared." (see [example](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_1_X_2021-09-19-0000+741dce/45602/triggerResults/23234.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D49_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D49+RecoFullGlobal_2026D49+HARVESTFullGlobal_2026D49/HLT.log)). Otoh, one can see that the new Phase-2 paths are there in the relevant wfs (see ["skipped triggers" here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_1_X_2021-09-19-0000+741dce/45602/triggerResults/23234.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D49_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D49+RecoFullGlobal_2026D49+HARVESTFullGlobal_2026D49/HLT.json)); maybe in the future we can try to improve a bit how `hltDiff` reports diffs;

* I personally don't know yet what is causing the [differences in wf 20434.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_1_X_2021-09-19-0000+741dce/45602/dqm-histo-comparison-summary.html).

cmsbuild commented:
Pull request #32903 was updated. @cmsbuild, @jordan-martins, @missirol, @bbilin, @wajidalikhan, @Martin-Grunewald, @AdrianoDee, @srimanob, @kskovpen can you please check and sign again.

missirol commented:
(Just a note for completeness)

As briefly discussed with Andrea offline, the previous differences in wf 20434.0 (which uses D41) might have been related to the HLT overwriting some (ES) modules with D49 settings, and these modules being used in DIGI sequences (since the DIGI and HLT steps are run together in that wf).

Now the use of the HLT Phase-2 menu is being restricted to the wf that uses the D49 geometry (i.e. the only geometry for which the Phase-2 HLT guarantees meaningful outputs).

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-741dce/18793/summary.html
**COMMIT**: 205db6811ff52d5745ce192cbf4a5090e1706596
**CMSSW**: CMSSW_11_1_X_2021-09-19-0000/slc7_amd64_gcc820
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32903/18793/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



The workflows 140.53 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 1275 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2784828
 - DQMHistoTests: Total failures: 2030
 - DQMHistoTests: Total nulls: 44
 - DQMHistoTests: Total successes: 2782704
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -45.703 KiB( 35 files compared)
 - DQMHistoSizes: changed ( 140.53 ): -44.531 KiB      Hcal/DigiRunHarvesting
 - DQMHistoSizes: changed ( 140.53 ): -1.172 KiB       RPC/DCSInfo
 - Checked 152 log files, 31 edm output root files, 36 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_1_X_2021-09-19-0000+741dce/45626/triggerResults): no differences found

Martin-Grunewald commented:
+1

srimanob commented:
Hi @fwyzard @missirol 

Do you have clue why we see difference in comparison, mostly in Physics sub-folder, e.g.
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_1_X_2021-09-19-0000+741dce/45626/23234.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D49_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D49+RecoFullGlobal_2026D49+HARVESTFullGlobal_2026D49/Physics_Exotica_Dijets_ak8PFJetsPuppi.html

For the problem of overwritten, is it specific for D41 only?

Thanks.

missirol commented:
Hi @srimanob

* for wf 23234.0, see [this comment](https://github.com/cms-sw/cmssw/pull/32903#issuecomment-923318693) from this PR;

* in short, it'd affect any non-D49 wf. The 'overwriting' is what I guess is happening; if it happens, it could explain differences in outputs for non-D49 scenarios where DIGI and HLT are run in the same step; the overwriting itself would also occur in D49 wfs, but it would lead to no changes in the configuration b/c the HLT modules are tailored towards D49.

srimanob commented:
+Upgrade

This PR implements Phase-2 HLT menu to 11_1 release. Currently, we should focus on D49 result which give meaningful of HLT result.

kskovpen commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_11_1_X IBs (tests are also fine) and once validation in the development release cycle CMSSW_12_1_X is complete. This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1
